File: LeetCode/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        LinkedList<TreeNode> st = new LinkedList<>();
        TreeNode cur = root;
        TreeNode small = null;
        while (cur != null || !st.isEmpty()) {
            while (cur != null) {
                st.push(cur);
                cur = cur.left;
            }
            cur = st.poll();
            if (small != null && cur.val <= small.val) return false;
            small = cur;
            cur = cur.right;
        }
        return true;
    }
} //TC: O(n), SC: O(height of tree), where n is node count

File: LeetCode/98. Validate Binary Search Tree.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isValidBST(r: TreeNode?): Boolean {
        var curr = r
        curr?.let {
            val st = Stack<TreeNode>()
            var prev: TreeNode? = null

            while (curr != null || st.isNotEmpty()) {
                while (curr != null) { //while the curr node is not null, push curr node and its left node to the stack
                    st.push(curr)
                    curr = curr!!.left
                }
                curr = st.pop() //pop when root is null
                if (prev != null && curr!!.`val` <= prev.`val`) return false //when current value is less or equal to previous one, return false
                prev = curr //set as previously visited node
                curr = curr!!.right //switch curr node to right node
            }
        }
        return true
    }
}
